cirq/sim/wave_function.py

```diff
@@ -26,6 +26,75 @@
     from cirq.sim import simulator
 
 
+def bloch_vector(state: Sequence, index: int) -> np.ndarray:
+    """Returns the bloch vector of a qubit.
+
+    Calculates the bloch vector of the qubit at index
+    in the wavefunction given by state.
+
+    Args:
+        state: A sequence representing a wave function in which
+            the ordering mapping to qubits follows the standard Kronecker
+            convention of numpy.kron.
+        index: index of qubit who's bloch vector we want to find.
+
+    Returns:
+        A length 3 numpy array representing the qubit's bloch vector.
+    """
+    rho = density_matrix(state, [index])
+    v = np.zeros(3, dtype=np.complex64)
+    v[0] = rho[1][0] + rho[0][1]
+    v[1] = 1.0j * (rho[0][1] - rho[1][0])
+    v[2] = rho[0][0] - rho[1][1]
+
+    return v
+
+
+def density_matrix(state: Sequence, indices: List[int] = None) -> np.ndarray:
+    """Returns the density matrix of the wavefunction.
+
+    Calculate the density matrix for the system on the given qubit
+    indices, with the qubits not in indices that are present in state
+    traced out. If indices is None the full density matrix for state
+    is returned.
+    For example:
+        state = np.array([1/np.sqrt(2), 1/np.sqrt(2)], dtype=np.complex64)
+        indices = None
+        gives us \rho = \begin{bmatrix}
+                            0.5 & 0.5
+                            0.5 & 0.5
+                        \end{bmatrix}
+
+    Args:
+        state: A sequence representing a wave function in which
+            the ordering mapping to qubits follows the standard Kronecker
+            convention of numpy.kron.
+        indices: list containing indices for qubits that you would like
+            to include in the density matrix (i.e.) qubits that WON'T
+            be traced out.
+
+    Returns:
+        A numpy array representing the density matrix.
+    """
+    n_qubits = _validate_num_qubits(state)
+    if indices is None:
+        return np.outer(np.conj(state), state)
+
+    all_indices = [i for i in range(n_qubits)]
+    _validate_indices(n_qubits, indices)
+
+    to_trace_out = sorted(list(set(all_indices).difference(indices)))
+    rho = np.outer(np.conj(state), state)
+
+    num_traced = 0
+    for index in to_trace_out:
```

Comment:
I wonder if np.einsum can do this faster.

Reply:
The above np.einsum was a fair bit faster than my original logic.